### DISCUSSION TITLE
decode JoiningProcessIdentificationDataType

### DISCUSSION THREAD

**Discussion by francy328** [eclipse-milo/milo#1280] [General]
Hi all, Kevin
I was able to call the method AbortJoiningProcess, I have writen the exactely argument, and store the codec;
But, when I call the method with UaExpert:
![image](https://github.com/eclipse/milo/assets/23223131/b732bf6a-81c0-4ca7-aa57-c148e112d631)

and have selected all of three field and put a value on all (field) like in figure, I can get a print of the value I have inserted.

If I just eliminate one filed as in figure:
![image](https://github.com/eclipse/milo/assets/23223131/f9b60083-8912-4437-926a-8d1f3eb14374)

and make a call, I get:
`18:28:16.664 [milo-shared-thread-pool-9] ERROR o.e.m.e.m.opcua.ModeledNamespace - Uncaught Throwable invoking method handler for methodId=NodeId{ns=8, id=92}.
org.eclipse.milo.opcua.stack.core.UaSerializationException: string length exceeds max message size (length=825372928, max=2097152)
	at org.eclipse.milo.opcua.stack.core.serialization.OpcUaBinaryStreamDecoder.readLengthPrefixedString(OpcUaBinaryStreamDecoder.java:434)
	at org.eclipse.milo.opcua.stack.core.serialization.OpcUaBinaryStreamDecoder.readString(OpcUaBinaryStreamDecoder.java:370)
	at org.eclipse.milo.opcua.stack.core.serialization.OpcUaBinaryStreamDecoder.readString(OpcUaBinaryStreamDecoder.java:592)
	at org.eclipse.milo.opcua.stack.core.types.structured.JoiningProcessIdentificationDataType$Codec.decode(JoiningProcessIdentificationDataType.java:92)
	at org.eclipse.milo.opcua.stack.core.types.structured.JoiningProcessIdentificationDataType$Codec.decode(JoiningProcessIdentificationDataType.java:1)
	at org.eclipse.milo.opcua.stack.core.serialization.codecs.GenericDataTypeCodec$GenericBinaryDataTypeCodec.decode(GenericDataTypeCodec.java:47)
	at org.eclipse.milo.opcua.stack.core.types.OpcUaDefaultBinaryEncoding.decode(OpcUaDefaultBinaryEncoding.java:117)
	at org.eclipse.milo.opcua.stack.core.types.builtin.ExtensionObject.lambda$0(ExtensionObject.java:105)
	at org.eclipse.milo.opcua.stack.core.util.Lazy.maybeCompute(Lazy.java:50)`

How that can be possible? It seams magic.

To get the decoded object I use:

```
ExtensionObject xo = (ExtensionObject) inputValues[1].getValue();
		 JoiningProcessIdentificationDataType decoded = (JoiningProcessIdentificationDataType) xo.decode(invocationContext.getServer().getSerializationContext());

```

where the class JoiningProcessIdentificationDataType is 

`package org.eclipse.milo.opcua.stack.core.types.structured;

import lombok.EqualsAndHashCode;
import lombok.ToString;
import lombok.experimental.SuperBuilder;

import org.eclipse.milo.opcua.stack.core.UaSerializationException;
import org.eclipse.milo.opcua.stack.core.serialization.SerializationContext;
import org.eclipse.milo.opcua.stack.core.serialization.UaDecoder;
import org.eclipse.milo.opcua.stack.core.serialization.UaEncoder;
import org.eclipse.milo.opcua.stack.core.serialization.UaStructure;
import org.eclipse.milo.opcua.stack.core.serialization.codecs.GenericDataTypeCodec;
import org.eclipse.milo.opcua.stack.core.types.builtin.ByteString;
import org.eclipse.milo.opcua.stack.core.types.builtin.DataValue;
import org.eclipse.milo.opcua.stack.core.types.builtin.ExpandedNodeId;
import org.eclipse.milo.opcua.stack.core.types.builtin.unsigned.UInteger;

@EqualsAndHashCode(
	    callSuper = false
	)
	@SuperBuilder(
	    toBuilder = true
	)

@ToString
public class JoiningProcessIdentificationDataType extends Structure implements UaStructure {
	
	 public static final ExpandedNodeId TYPE_ID = ExpandedNodeId.parse("nsu=http://opcfoundation.org/UA/Machinery/;i=5121");

	 public static final ExpandedNodeId XML_ENCODING_ID = ExpandedNodeId.parse("nsu=http://opcfoundation.org/UA/;i=12676");

	 public static final ExpandedNodeId BINARY_ENCODING_ID = ExpandedNodeId.parse("nsu=http://opcfoundation.org/UA/Machinery/;i=5121");

	 
	 private final String JoiningProcessId;
	 private final String JoiningProcessOriginId;
	 private final String SelectionName;
	 

	public JoiningProcessIdentificationDataType(String joiningProcessId, String joiningProcessOriginId,
			String selectionName) {
			
			JoiningProcessId = joiningProcessId;
			JoiningProcessOriginId = joiningProcessOriginId;
			SelectionName = selectionName;
		}
	 
	  @Override
	    public ExpandedNodeId getTypeId() {
	        return TYPE_ID;
	    }

	    @Override
	    public ExpandedNodeId getXmlEncodingId() {
	        return XML_ENCODING_ID;
	    }

	    @Override
	    public ExpandedNodeId getBinaryEncodingId() {
	        return BINARY_ENCODING_ID;
	    }
	    

		public String getJoiningProcessId() {
			return JoiningProcessId;
		}

		public String getJoiningProcessOriginId() {
			return JoiningProcessOriginId;
		}

		public String getSelectionName() {
			return SelectionName;
		} 
	 
	 
	 
	    public static final class Codec extends GenericDataTypeCodec<JoiningProcessIdentificationDataType> {
	        @Override
	        public Class<JoiningProcessIdentificationDataType> getType() {
	            return JoiningProcessIdentificationDataType.class;
	        }

	        @Override
	        public JoiningProcessIdentificationDataType decode(SerializationContext context, UaDecoder decoder) {
	        	System.out.println("decode JoiningProcessIdentificationDataType ");
	        
	        	String JoiningProcessId = decoder.readString("JoiningProcessId");
	        	System.out.println("JoiningProcessId "+JoiningProcessId);
	        	String JoiningProcessOriginId = decoder.readString("JoiningProcessOriginId");
	        	System.out.println("JoiningProcessOriginId "+JoiningProcessOriginId);
	        	String SelectionName = decoder.readString("SelectionName");
	        	System.out.println("SelectionName "+SelectionName);
       	
	        	return new JoiningProcessIdentificationDataType(JoiningProcessId,JoiningProcessOriginId,SelectionName);
	        }

			@Override
			public void encode(SerializationContext context, UaEncoder writer,
					JoiningProcessIdentificationDataType value) throws UaSerializationException {
				
				System.out.println("encode JoiningProcessIdentificationDataType ");
				writer.writeString("JoiningProcessId", value.getJoiningProcessId());
				writer.writeString("JoiningProcessOriginId", value.getJoiningProcessOriginId());
				writer.writeString("SelectionName", value.getSelectionName());	
			}
	    }
}
`

**Comment by kevinherron** [comment]
The codec isn't implemented correctly. If you look at the definition of `JoiningProcessIdentificationDataType` in the NodeSet2.xml file  for this information model, or the [definition](https://reference.opcfoundation.org/IJT/Base/v100/docs/10.21#_Ref129857566), you'll see the fields are all optional, which means this structure is encoded according to the rules for a [Structure with optional fields](https://reference.opcfoundation.org/Core/Part6/v105/docs/5.2.7).

**Comment by francy328** [comment]
I've read  the documentation, I also try to debug looking at the code, but I'm still in trouble, no way to me so far to understand what is wrong; I've changed the following:

> public static final ExpandedNodeId TYPE_ID = ExpandedNodeId.parse("nsu=http://opcfoundation.org/UA/IJT/Base/;i=3029");

	 public static final ExpandedNodeId XML_ENCODING_ID = ExpandedNodeId.parse("nsu=http://opcfoundation.org/UA/;i=12676");

	 public static final ExpandedNodeId BINARY_ENCODING_ID = ExpandedNodeId.parse("nsu=http://opcfoundation.org/UA/IJT/Base/;i=5121");

And this is the way I registered the code in the dictionary:

`        NodeId nodeId = new NodeId(6, 5121);

        binaryDictionary.registerStructCodec(
                new JoiningProcessIdentificationDataType.Codec().asBinaryCodec(),
                "JoiningProcessIdentificationDataType",
                nodeId,
                nodeId
            );`

I cannot use the usaul format like in:
`binaryDictionary.registerStructCodec(
            new Argument.Codec().asBinaryCodec(),
            "Argument",
            Argument.TYPE_ID.toNodeId(namespaceTable).orElseThrow(
                () -> new IllegalStateException("namespace not registered: " + NAMESPACE_URI)),
            Argument.BINARY_ENCODING_ID.toNodeId(namespaceTable).orElseThrow(
                () -> new IllegalStateException("namespace not registered: " + NAMESPACE_URI))
        );`


because the namespaceTable I need having an index of 6 (otherwise the program complain that that encoding is missing).

In the class JoiningProcessIdentificationDataType, i also have changed the type of JoiningProcessId from String to ByteString but still not working, 

How I can understand the right way of implementing the codec? What wrong with me?

**Reply by kevinherron** [reply]
What's not working? what error are you getting or what's wrong?

> because the namespaceTable I need having an index of 6 (otherwise the program complain that that encoding is missing).

what do you mean by this? Have you added any/all namespace URIs to the server's namespace table?

**Reply by francy328** [reply]
The error is the same as indicated in the answer, if I do not select all three field I get : 
`18:28:16.664 [milo-shared-thread-pool-9] ERROR o.e.m.e.m.opcua.ModeledNamespace - Uncaught Throwable invoking method handler for methodId=NodeId{ns=8, id=92}. org.eclipse.milo.opcua.stack.core.UaSerializationException: string length exceeds max message size (length=825372928, max=2097152) at `

I mean that to register the codec I dit like this:

`NodeId nodeId = new NodeId(6, 5121);

    binaryDictionary.registerStructCodec(
            new JoiningProcessIdentificationDataType.Codec().asBinaryCodec(),
            "JoiningProcessIdentificationDataType",
            nodeId,
            nodeId
        );``


Do you mean to add the namespace in the following class
![image](https://github.com/user-attachments/assets/fe49b648-daa5-4f3c-b437-d2a6148d0b76)

in other word, something like this:
uriTable.put(ushort(6), Namespaces.OPC_JOIN);

**Reply by kevinherron** [reply]
Yes, register your namespaces, the OpcUaServer has one via `getNamespaceTable`.

Are you sure you fixed your codec implementation? What does it look like now?

**Reply by francy328** [reply]
In attached (renamed .log, in order to upload) what I think is my codec implementation; No I not sure I ve fixed because I still do not understand how to do that :) 
[JoiningProcessIdentificationDataType.java.log](https://github.com/user-attachments/files/16256193/JoiningProcessIdentificationDataType.java.log)

I will register the namespace via getNamespaceTable and I will let you know.

I also find an error when I call 
binaryDictionary.registerStructCodec
as argument I pass two times the same nodeId, while in the in the rest of your code, they differ!
